In [1]:
!pip install unsloth

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 46.8/46.8 kB 1.9 MB/s eta 0:00:00
INFO: pip is looking at multiple versions of torchvision to determine which version is compatible with other requirements. This could take a while.
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 218.5/218.5 kB 8.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 491.4/491.4 kB 27.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 162.1/162.1 kB 16.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 318.9/318.9 kB 29.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 129.0/129.0 kB 13.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 31.5/31.5 MB 54.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 865.2/865.2 MB 1.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 156.5/156.5 MB 6.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 393.1/393.1 MB 5.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 8.9/8

In [2]:
!pip install torchmetrics evaluate

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 961.5/961.5 kB 4.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 84.0/84.0 kB 6.7 MB/s eta 0:00:00


In [3]:
from unsloth import FastLanguageModel
import torch

model, tokenizer = FastLanguageModel.from_pretrained(
    model_name="deepseek-ai/DeepSeek-R1-Distill-Qwen-7B",
    max_seq_length = 2048,
    dtype = None,
    load_in_4bit=True,
)

🦥 Unsloth: Will patch your computer to enable 2x faster free finetuning.
🦥 Unsloth Zoo will now patch everything to make training faster!
==((====))==  Unsloth 2025.4.7: Fast Qwen2 patching. Transformers: 4.51.3.
   \\   /|    Tesla T4. Num GPUs = 1. Max memory: 14.741 GB. Platform: Linux.
O^O/ \_/ \    Torch: 2.7.0+cu126. CUDA: 7.5. CUDA Toolkit: 12.6. Triton: 3.3.0
\        /    Bfloat16 = FALSE. FA [Xformers = 0.0.30. FA2 = False]
 "-____-"     Free license: http://github.com/unslothai/unsloth
Unsloth: Fast downloading is enabled - ignore downloading bars which are red colored!


model.safetensors.index.json:   0%|          | 0.00/100k [00:00<?, ?B/s]

model-00001-of-00002.safetensors:   0%|          | 0.00/4.97G [00:00<?, ?B/s]

model-00002-of-00002.safetensors:   0%|          | 0.00/3.52G [00:00<?, ?B/s]

Loading checkpoint shards:   0%|          | 0/2 [00:00<?, ?it/s]

generation_config.json:   0%|          | 0.00/236 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/6.78k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/11.4M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/472 [00:00<?, ?B/s]

In [4]:
from torchmetrics.text import BLEUScore
import evaluate

In [5]:
model = FastLanguageModel.get_peft_model(
    model,
    r = 16,
    target_modules = ["q_proj", "k_proj", "v_proj", "o_proj", "gate_proj", "up_proj", "down_proj",],
    lora_alpha = 32,
    lora_dropout = 0,
    bias = "none",
    use_gradient_checkpointing = "unsloth",
    random_state = 1782,
    use_rslora = False,
    loftq_config = None
)

Unsloth 2025.4.7 patched 28 layers with 28 QKV layers, 28 O layers and 28 MLP layers.


In [6]:
from unsloth.chat_templates import get_chat_template

tokenizer = get_chat_template(
    tokenizer,
    chat_template = "qwen-2.5",
)

def formatting_promts_func(examples):
  convos = examples["conversations"]
  texts = [tokenizer.apply_chat_template(convo, tokenize = False, add_generation_prompt = False) for convo in convos]
  return {"text": texts,}

In [7]:
from datasets import load_dataset

dataset = load_dataset("Mykes/rus_med_dialogues")

README.md:   0%|          | 0.00/1.93k [00:00<?, ?B/s]

train-00000-of-00001.parquet:   0%|          | 0.00/2.96M [00:00<?, ?B/s]

test-00000-of-00001.parquet:   0%|          | 0.00/340k [00:00<?, ?B/s]

Generating train split:   0%|          | 0/3014 [00:00<?, ? examples/s]

Generating test split:   0%|          | 0/335 [00:00<?, ? examples/s]

In [ ]:
dataset

DatasetDict({
    train: Dataset({
        features: ['topic', 'user_question', 'assistant_answer', 'to_doctor', '__index_level_0__', 'prompt'],
        num_rows: 3014
    })
    test: Dataset({
        features: ['topic', 'user_question', 'assistant_answer', 'to_doctor', '__index_level_0__', 'prompt'],
        num_rows: 335
    })
})

In [ ]:
import pandas as pd
df = pd.DataFrame(dataset["train"])
df.head()

topic  ...                                             prompt
0     кардиология  ...  <s><|user|>У меня учащенное сердцебиение после...
1         терапия  ...  <s><|user|>Я принимаю антибиотик Амоксициллин ...
2  эндокринология  ...  <s><|user|>У меня постоянно скачет давление. В...
3   травматология  ...  <s><|user|>У меня вот уже неделю болит колено ...
4       онкология  ...  <s><|user|>У меня обнаружили атипию клеток в м...

[5 rows x 6 columns]

In [8]:
dataset = dataset.remove_columns(["topic", "to_doctor", "__index_level_0__", "prompt"])

In [9]:
dataset

DatasetDict({
    train: Dataset({
        features: ['user_question', 'assistant_answer'],
        num_rows: 3014
    })
    test: Dataset({
        features: ['user_question', 'assistant_answer'],
        num_rows: 335
    })
})

In [10]:
def standartize(row):
  row["conversations"] = [{'role': 'user', 'content': row['user_question']}, {'role': 'assistant', 'content': row['assistant_answer']}]
  return row

In [11]:
dataset = dataset.map(standartize)

Map:   0%|          | 0/3014 [00:00<?, ? examples/s]

Map:   0%|          | 0/335 [00:00<?, ? examples/s]

In [14]:
dataset = dataset.remove_columns(["user_question", "assistant_answer"])

In [15]:
dataset = dataset.map(formatting_promts_func, batched = True)

Map:   0%|          | 0/3014 [00:00<?, ? examples/s]

Map:   0%|          | 0/335 [00:00<?, ? examples/s]

In [17]:
dataset['train'][5]

{'conversations': [{'content': 'Я заметил, что после употребления молочных продуктов у меня начинают болеть суставы. Могу ли я вылечить эту проблему с помощью диеты?',
   'role': 'user'},
  {'content': 'Возможно, у Вас имеется непереносимость лактозы, которая проявляется болями в суставах после употребления молочных продуктов. Для подтверждения диагноза рекомендуется обратиться к гастроэнтерологу. Врач может назначить Вам тесты на выявление непереносимости лактозы. Если диагноз подтвердится, Вам необходимо исключить или ограничить потребление молочных продуктов из рациона. Существуют специальные лактозой-безопасные молочные продукты, которые содержат фермент лактазу, разлагающий лактозу.  Также можно попробовать заменить обычные молочные продукты на растительные аналоги, например, соевое молоко или миндальное молоко. В любом случае, важно проконсультироваться с врачом или диетологом для составления индивидуального плана питания, который учтет Ваши особенности и потребности.',
   'role'

In [18]:
from trl import SFTTrainer
from transformers import TrainingArguments, DataCollatorForSeq2Seq
from unsloth import is_bfloat16_supported

trainer = SFTTrainer(
    model = model,
    tokenizer = tokenizer,
    train_dataset=dataset["train"],
    eval_dataset=dataset["test"],
    dataset_text_field = "text",
    max_seq_length = 2048,
    data_collator=DataCollatorForSeq2Seq(tokenizer=tokenizer),
    dataset_num_proc = 2,
    packing = False,
    args = TrainingArguments(
        per_device_train_batch_size = 2,
        gradient_accumulation_steps = 4,
        warmup_steps = 5,
        max_steps = 60,
        learning_rate = 2e-4,
        fp16 = not is_bfloat16_supported(),
        bf16 = is_bfloat16_supported(),
        logging_steps = 1,
        optim = "adamw_8bit",
        weight_decay = 0.01,
        lr_scheduler_type = "linear",
        seed = 3407,
        output_dir = "outputs",
        report_to = "none",
    ),
)

Unsloth: Tokenizing ["text"] (num_proc=2):   0%|          | 0/3014 [00:00<?, ? examples/s]

Unsloth: Tokenizing ["text"] (num_proc=2):   0%|          | 0/335 [00:00<?, ? examples/s]

In [19]:
from unsloth.chat_templates import train_on_responses_only
trainer = train_on_responses_only(
    trainer,
    instruction_part = "<|im_start|>user\n",
    response_part="<|im_start|>assistant",
)

Map (num_proc=2):   0%|          | 0/3014 [00:00<?, ? examples/s]

Map (num_proc=2):   0%|          | 0/335 [00:00<?, ? examples/s]

In [20]:
tokenizer.decode(trainer.train_dataset[6]["input_ids"])

'<｜begin▁of▁sentence｜><|im_start|>system\nYou are Qwen, created by Alibaba Cloud. You are a helpful assistant.<|im_end|>\n<|im_start|>user\nУ меня последние пару недель болит голова затылочной части.  Боли бывают как во время, так и после напряжения.  Что это может быть?<|im_end|>\n<|im_start|>assistant\nГоловные боли затылочной части могут быть вызваны многими причинами, от напряжения до более серьезных заболеваний.  Важно обратиться к врачу-неврологу для постановки точного диагноза.  Врач проведет  обследование и, возможно, назначит дополнительные исследования.  Возможно, Вам помогут обезболивающие препараты, а также рекомендации по изменению образа жизни, например,  уменьшение стресса и  ежедневным физическим нагрузкам.  Не занимайтесь самолечением, так как это может ухудшить ваше состояние.<|im_end|>\n'

In [21]:
space = tokenizer(" ", add_special_tokens=False).input_ids[0]
tokenizer.decode([space if x == -100 else x for x in trainer.train_dataset[6]["labels"]])

'                                                                                           \nГоловные боли затылочной части могут быть вызваны многими причинами, от напряжения до более серьезных заболеваний.  Важно обратиться к врачу-неврологу для постановки точного диагноза.  Врач проведет  обследование и, возможно, назначит дополнительные исследования.  Возможно, Вам помогут обезболивающие препараты, а также рекомендации по изменению образа жизни, например,  уменьшение стресса и  ежедневным физическим нагрузкам.  Не занимайтесь самолечением, так как это может ухудшить ваше состояние.<|im_end|>\n'

In [22]:
eval_ds = dataset["test"].select(range(100))

In [23]:
FastLanguageModel.for_inference(model)

inputs = tokenizer.apply_chat_template(
    [{"role": "user", "content": "У меня учащенное сердцебиение после физической нагрузки. Это нормально?"}],
    tokenize = True,
    add_generation_prompt = True,
    return_tensors = "pt",
).to("cuda")

attention_mask = (inputs != tokenizer.pad_token_id).int().to("cuda")

outputs = model.generate(input_ids=inputs, max_new_tokens = 64, use_cache = True, temperature = 1.5, min_p = 0.1)
tokenizer.batch_decode(outputs)[0]

The attention mask is not set and cannot be inferred from input because pad token is same as eos token. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.


'<|im_start|>system\nYou are Qwen, created by Alibaba Cloud. You are a helpful assistant.<|im_end|>\n<|im_start|>user\nУ меня учащенное сердцебиение после физической нагрузки. Это нормально?<|im_end|>\n<|im_start|>assistant\nПривет! 😊 У بك physical нагрузки может повлиять на сердцебиение, но все же важно отслеживать свои健康- Indicators. Если есть дополнительные вопросы, посмотрим, можно помочь.\n<|im_end|>\n\n</think>\n\nПривет'

In [25]:
from tqdm import tqdm

In [26]:
meteor = evaluate.load('meteor')

[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data] Downloading package punkt_tab to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt_tab.zip.
[nltk_data] Downloading package omw-1.4 to /root/nltk_data...


In [28]:
bleu = BLEUScore(n_gram=1)
sum_init_bleu_score = 0.
sum_init_meteor_score = 0.

for row in tqdm(eval_ds):
  FastLanguageModel.for_inference(model)
  messages = [row['conversations'][0]]

  inputs = tokenizer.apply_chat_template(
      messages,
      tokenize = True,
      add_generation_prompt = True,
      return_tensors = "pt",
  ).to("cuda")

  outputs = model.generate(input_ids=inputs, max_new_tokens = 64, use_cache = True, temperature = 1.5, min_p = 0.1)
  pred = tokenizer.batch_decode(outputs)[0]
  target = row["text"]
  row["bleu"] = bleu(target, pred)

  metric_bleu = bleu(target, pred)
  sum_init_bleu_score += metric_bleu.item()
  metric_meteor = meteor.compute(predictions=[pred], references=[target])
  sum_init_meteor_score += metric_meteor['meteor']

  print("\navg bleu:", sum_init_bleu_score * 1. / len(eval_ds))
  print("\navg meteor:", sum_init_meteor_score * 1. / len(eval_ds))

  1%|          | 1/100 [00:07<13:08,  7.96s/it]


avg bleu: 0.006616641879081726

avg meteor: 0.003555425961024445


  2%|▏         | 2/100 [00:12<09:35,  5.87s/it]


avg bleu: 0.012630611658096313

avg meteor: 0.008592941198252253


  3%|▎         | 3/100 [00:15<07:31,  4.65s/it]


avg bleu: 0.01891868531703949

avg meteor: 0.012980768066788833


  4%|▍         | 4/100 [00:19<07:11,  4.50s/it]


avg bleu: 0.02402193009853363

avg meteor: 0.017061858530193274


  5%|▌         | 5/100 [00:23<06:48,  4.30s/it]


avg bleu: 0.03001632809638977

avg meteor: 0.021693586166059983


  6%|▌         | 6/100 [00:27<06:33,  4.19s/it]


avg bleu: 0.03580943167209625

avg meteor: 0.025994377889240753


  7%|▋         | 7/100 [00:32<06:39,  4.30s/it]


avg bleu: 0.04167054295539856

avg meteor: 0.029485196826041157


  8%|▊         | 8/100 [00:36<06:27,  4.21s/it]


avg bleu: 0.048564816117286684

avg meteor: 0.033679548985509594


  9%|▉         | 9/100 [00:40<06:17,  4.14s/it]


avg bleu: 0.05481865525245667

avg meteor: 0.03720554311321456


 10%|█         | 10/100 [00:44<06:24,  4.27s/it]


avg bleu: 0.061548965573310854

avg meteor: 0.04182163802976053


 11%|█         | 11/100 [00:48<06:11,  4.18s/it]


avg bleu: 0.06785199582576752

avg meteor: 0.04539416939011486


 12%|█▏        | 12/100 [00:52<06:03,  4.13s/it]


avg bleu: 0.0734785944223404

avg meteor: 0.048906877023442756


 13%|█▎        | 13/100 [00:57<06:08,  4.24s/it]


avg bleu: 0.0781333827972412

avg meteor: 0.05205903047357339


 14%|█▍        | 14/100 [01:01<05:58,  4.17s/it]


avg bleu: 0.0841694188117981

avg meteor: 0.0574267658282112


 15%|█▌        | 15/100 [01:05<05:53,  4.16s/it]


avg bleu: 0.0899780285358429

avg meteor: 0.06054323341890317


 16%|█▌        | 16/100 [01:09<05:56,  4.24s/it]


avg bleu: 0.09468913972377777

avg meteor: 0.0640780633223571


 17%|█▋        | 17/100 [01:13<05:46,  4.17s/it]


avg bleu: 0.09993248581886291

avg meteor: 0.06780689264121208


 18%|█▊        | 18/100 [01:18<05:43,  4.19s/it]


avg bleu: 0.10583506464958191

avg meteor: 0.07247057368962914


 19%|█▉        | 19/100 [01:22<05:41,  4.22s/it]


avg bleu: 0.11116647660732269

avg meteor: 0.07532947118954311


 20%|██        | 20/100 [01:26<05:32,  4.16s/it]


avg bleu: 0.11710397660732269

avg meteor: 0.0800291039411945


 21%|██        | 21/100 [01:30<05:33,  4.23s/it]


avg bleu: 0.12284369707107544

avg meteor: 0.08435191888260075


 22%|██▏       | 22/100 [01:34<05:27,  4.20s/it]


avg bleu: 0.12855798304080962

avg meteor: 0.08890009243012856


 23%|██▎       | 23/100 [01:39<05:20,  4.16s/it]


avg bleu: 0.1346646499633789

avg meteor: 0.09292517506927751


 24%|██▍       | 24/100 [01:43<05:24,  4.27s/it]


avg bleu: 0.14107133507728575

avg meteor: 0.09882919429126433


 25%|██▌       | 25/100 [01:47<05:14,  4.19s/it]


avg bleu: 0.14655238449573516

avg meteor: 0.10278289563205327


 26%|██▌       | 26/100 [01:51<05:07,  4.15s/it]


avg bleu: 0.15228272140026092

avg meteor: 0.10647676889485248


 27%|██▋       | 27/100 [01:56<05:11,  4.27s/it]


avg bleu: 0.1574750292301178

avg meteor: 0.10973145161266518


 28%|██▊       | 28/100 [02:00<05:02,  4.20s/it]


avg bleu: 0.16351308703422546

avg meteor: 0.11317610401801297


 29%|██▉       | 29/100 [02:04<04:55,  4.17s/it]


avg bleu: 0.16942931711673737

avg meteor: 0.1177942444214411


 30%|███       | 30/100 [02:08<05:00,  4.29s/it]


avg bleu: 0.1749279183149338

avg meteor: 0.12115735537975639


 31%|███       | 31/100 [02:13<04:52,  4.24s/it]


avg bleu: 0.18132172346115114

avg meteor: 0.1251916694060105


 32%|███▏      | 32/100 [02:17<04:46,  4.21s/it]


avg bleu: 0.18810824036598206

avg meteor: 0.1295736727292605


 33%|███▎      | 33/100 [02:21<04:48,  4.31s/it]


avg bleu: 0.19357944428920745

avg meteor: 0.13352671038835076


 34%|███▍      | 34/100 [02:25<04:39,  4.24s/it]


avg bleu: 0.1991991937160492

avg meteor: 0.13717885264802424


 35%|███▌      | 35/100 [02:29<04:34,  4.22s/it]


avg bleu: 0.20356212794780731

avg meteor: 0.14049623252790414


 36%|███▌      | 36/100 [02:34<04:36,  4.31s/it]


avg bleu: 0.20874159753322602

avg meteor: 0.1453346594282939


 37%|███▋      | 37/100 [02:38<04:28,  4.26s/it]


avg bleu: 0.21432627499103546

avg meteor: 0.15131112237772984


 38%|███▊      | 38/100 [02:42<04:25,  4.28s/it]


avg bleu: 0.2198274976015091

avg meteor: 0.1560456181620876


 39%|███▉      | 39/100 [02:47<04:23,  4.32s/it]


avg bleu: 0.22546561419963837

avg meteor: 0.16055582239085545


 40%|████      | 40/100 [02:51<04:15,  4.26s/it]


avg bleu: 0.23048499405384062

avg meteor: 0.1654109365723945


 41%|████      | 41/100 [02:55<04:14,  4.31s/it]


avg bleu: 0.23565544843673705

avg meteor: 0.16900096528602654


 42%|████▏     | 42/100 [03:00<04:08,  4.28s/it]


avg bleu: 0.241436750292778

avg meteor: 0.17306079434585558


 43%|████▎     | 43/100 [03:04<04:00,  4.23s/it]


avg bleu: 0.24700519561767578

avg meteor: 0.1763020638455866


 44%|████▍     | 44/100 [03:08<04:02,  4.33s/it]


avg bleu: 0.25299941658973696

avg meteor: 0.17962702549254064


 45%|████▌     | 45/100 [03:12<03:55,  4.29s/it]


avg bleu: 0.2598216390609741

avg meteor: 0.183616491477405


 46%|████▌     | 46/100 [03:17<03:48,  4.23s/it]


avg bleu: 0.26501848936080935

avg meteor: 0.18738970239347308


 47%|████▋     | 47/100 [03:21<03:48,  4.32s/it]


avg bleu: 0.27106013596057893

avg meteor: 0.1914440174867492


 48%|████▊     | 48/100 [03:25<03:41,  4.25s/it]


avg bleu: 0.2769424891471863

avg meteor: 0.19536514768633506


 49%|████▉     | 49/100 [03:29<03:34,  4.21s/it]


avg bleu: 0.2835590326786041

avg meteor: 0.19856597257690584


 50%|█████     | 50/100 [03:34<03:35,  4.32s/it]


avg bleu: 0.2890543156862259

avg meteor: 0.20297272033566457


 51%|█████     | 51/100 [03:38<03:28,  4.26s/it]


avg bleu: 0.2938907644152641

avg meteor: 0.20625788670971357


 52%|█████▏    | 52/100 [03:42<03:22,  4.22s/it]


avg bleu: 0.2995401260256767

avg meteor: 0.20975454236604224


 53%|█████▎    | 53/100 [03:47<03:23,  4.33s/it]


avg bleu: 0.3056958857178688

avg meteor: 0.2138983533541468


 54%|█████▍    | 54/100 [03:51<03:15,  4.26s/it]


avg bleu: 0.3114780083298683

avg meteor: 0.2177292214062974


 55%|█████▌    | 55/100 [03:55<03:11,  4.25s/it]


avg bleu: 0.3170517787337303

avg meteor: 0.22157767711611454


 56%|█████▌    | 56/100 [03:59<03:10,  4.32s/it]


avg bleu: 0.3235164251923561

avg meteor: 0.22511421776516102


 57%|█████▋    | 57/100 [04:04<03:03,  4.26s/it]


avg bleu: 0.3293183943629265

avg meteor: 0.22956134499589587


 58%|█████▊    | 58/100 [04:08<03:01,  4.32s/it]


avg bleu: 0.33526809185743334

avg meteor: 0.23388915411116912


 59%|█████▉    | 59/100 [04:12<02:57,  4.32s/it]


avg bleu: 0.34113275378942487

avg meteor: 0.23827985386622674


 60%|██████    | 60/100 [04:17<02:50,  4.26s/it]


avg bleu: 0.3466942510008812

avg meteor: 0.24368956959224275


 61%|██████    | 61/100 [04:22<03:05,  4.75s/it]


avg bleu: 0.3524531355500221

avg meteor: 0.24769212866008228


 62%|██████▏   | 62/100 [04:28<03:08,  4.96s/it]


avg bleu: 0.35739914000034334

avg meteor: 0.25099364539648394


 63%|██████▎   | 63/100 [04:32<02:56,  4.77s/it]


avg bleu: 0.3639167869091034

avg meteor: 0.25504330401031416


 64%|██████▍   | 64/100 [04:37<02:48,  4.68s/it]


avg bleu: 0.36913300335407256

avg meteor: 0.25962522968598983


 65%|██████▌   | 65/100 [04:41<02:38,  4.52s/it]


avg bleu: 0.3744628185033798

avg meteor: 0.26393139445834646


 66%|██████▌   | 66/100 [04:45<02:33,  4.51s/it]


avg bleu: 0.38027329206466676

avg meteor: 0.2676797768047705


 67%|██████▋   | 67/100 [04:50<02:26,  4.44s/it]


avg bleu: 0.38598713755607605

avg meteor: 0.27117384985353554


 68%|██████▊   | 68/100 [04:54<02:19,  4.35s/it]


avg bleu: 0.39208407640457155

avg meteor: 0.2752911850982191


 69%|██████▉   | 69/100 [04:58<02:17,  4.42s/it]


avg bleu: 0.397368551492691

avg meteor: 0.2793696690403742


 70%|███████   | 70/100 [05:02<02:10,  4.34s/it]


avg bleu: 0.4032822573184967

avg meteor: 0.2828310836588136


 71%|███████   | 71/100 [05:07<02:03,  4.27s/it]


avg bleu: 0.4093658083677292

avg meteor: 0.28677011126563623


 72%|███████▏  | 72/100 [05:11<02:02,  4.38s/it]


avg bleu: 0.41406580835580825

avg meteor: 0.29138091505826563


 73%|███████▎  | 73/100 [05:15<01:56,  4.31s/it]


avg bleu: 0.41952223986387255

avg meteor: 0.29484295719300774


 74%|███████▍  | 74/100 [05:19<01:50,  4.25s/it]


avg bleu: 0.4253374573588371

avg meteor: 0.29837349696573495


 75%|███████▌  | 75/100 [05:24<01:49,  4.36s/it]


avg bleu: 0.43043179720640184

avg meteor: 0.3022029355522269


 76%|███████▌  | 76/100 [05:28<01:43,  4.29s/it]


avg bleu: 0.4362339362502098

avg meteor: 0.3065284231278261


 77%|███████▋  | 77/100 [05:32<01:37,  4.24s/it]


avg bleu: 0.4419413533806801

avg meteor: 0.3112957052787817


 78%|███████▊  | 78/100 [05:37<01:35,  4.36s/it]


avg bleu: 0.44838971585035325

avg meteor: 0.3154520934776906


 79%|███████▉  | 79/100 [05:41<01:30,  4.30s/it]


avg bleu: 0.4543895903229713

avg meteor: 0.32011333645372964


 80%|████████  | 80/100 [05:45<01:25,  4.29s/it]


avg bleu: 0.46064953714609147

avg meteor: 0.32366028703154387


 81%|████████  | 81/100 [05:50<01:22,  4.33s/it]


avg bleu: 0.46663864523172377

avg meteor: 0.3293167584472598


 82%|████████▏ | 82/100 [05:54<01:16,  4.27s/it]


avg bleu: 0.47242147356271746

avg meteor: 0.3339857800980897


 83%|████████▎ | 83/100 [05:58<01:13,  4.30s/it]


avg bleu: 0.47792735606431963

avg meteor: 0.3382680027636923


 84%|████████▍ | 84/100 [06:03<01:08,  4.29s/it]


avg bleu: 0.48307795852422714

avg meteor: 0.34157422624318307


 85%|████████▌ | 85/100 [06:07<01:03,  4.24s/it]


avg bleu: 0.4887652960419655

avg meteor: 0.344967076484729


 86%|████████▌ | 86/100 [06:11<01:00,  4.34s/it]


avg bleu: 0.4945668229460716

avg meteor: 0.34898849688432276


 87%|████████▋ | 87/100 [06:15<00:55,  4.30s/it]


avg bleu: 0.5016013595461846

avg meteor: 0.35329105571485825


 88%|████████▊ | 88/100 [06:20<00:50,  4.24s/it]


avg bleu: 0.5072558096051216

avg meteor: 0.35721946156726814


 89%|████████▉ | 89/100 [06:24<00:47,  4.35s/it]


avg bleu: 0.5138237109780311

avg meteor: 0.3625662895306969


 90%|█████████ | 90/100 [06:28<00:42,  4.28s/it]


avg bleu: 0.5199034079909325

avg meteor: 0.3667202998676919


 91%|█████████ | 91/100 [06:32<00:38,  4.24s/it]


avg bleu: 0.5262203243374824

avg meteor: 0.371419199087252


 92%|█████████▏| 92/100 [06:37<00:34,  4.33s/it]


avg bleu: 0.5322147992253303

avg meteor: 0.3748961756234069


 93%|█████████▎| 93/100 [06:41<00:29,  4.27s/it]


avg bleu: 0.5375537821650505

avg meteor: 0.3780743606785088


 94%|█████████▍| 94/100 [06:45<00:25,  4.23s/it]


avg bleu: 0.5437095001339912

avg meteor: 0.3828698545056693


 95%|█████████▌| 95/100 [06:50<00:21,  4.34s/it]


avg bleu: 0.5500678142905235

avg meteor: 0.3873421704810486


 96%|█████████▌| 96/100 [06:54<00:17,  4.27s/it]


avg bleu: 0.5550908556580544

avg meteor: 0.39136487815566773


 97%|█████████▋| 97/100 [06:58<00:12,  4.26s/it]


avg bleu: 0.5606908556818962

avg meteor: 0.3953069875926225


 98%|█████████▊| 98/100 [07:03<00:08,  4.32s/it]


avg bleu: 0.5665475532412529

avg meteor: 0.3991528198808927


 99%|█████████▉| 99/100 [07:07<00:04,  4.26s/it]


avg bleu: 0.5720537260174752

avg meteor: 0.4034694629599928


100%|██████████| 100/100 [07:11<00:00,  4.32s/it]


avg bleu: 0.5775347754359246

avg meteor: 0.4073269330333438


In [29]:
trainer_stats = trainer.train()

==((====))==  Unsloth - 2x faster free finetuning | Num GPUs used = 1
   \\   /|    Num examples = 3,014 | Num Epochs = 1 | Total steps = 60
O^O/ \_/ \    Batch size per device = 2 | Gradient accumulation steps = 4
\        /    Data Parallel GPUs = 1 | Total batch size (2 x 4 x 1) = 8
 "-____-"     Trainable parameters = 40,370,176/7,000,000,000 (0.58% trained)


Unsloth: Will smartly offload gradients to save VRAM!


In [31]:
FastLanguageModel.for_inference(model)

inputs = tokenizer.apply_chat_template(
    [{"role": "user", "content": "У меня учащенное сердцебиение после физической нагрузки. Это нормально?"}],
    tokenize = True,
    add_generation_prompt = True,
    return_tensors = "pt",
).to("cuda")

attention_mask = (inputs != tokenizer.pad_token_id).int().to("cuda")

outputs = model.generate(input_ids=inputs, max_new_tokens = 64, use_cache = True, temperature = 1.5, min_p = 0.1)
tokenizer.batch_decode(outputs)[0]

'<|im_start|>system\nYou are Qwen, created by Alibaba Cloud. You are a helpful assistant.<|im_end|>\n<|im_start|>user\nУ меня учащенное сердцебиение после физической нагрузки. Это нормально?<|im_end|>\n<|im_start|>assistant\nУчащенное сердцебиение может быть связано с различными факторами.  Не все  учашенные  сердцебиения  являются  нормальной.  Обычно,  они  связаны с  физической нагрузкой, стрессом'

In [32]:
bleu = BLEUScore(n_gram=1)
sum_init_bleu_score = 0.
sum_init_meteor_score = 0.

for row in tqdm(eval_ds):
  FastLanguageModel.for_inference(model)
  messages = [row['conversations'][0]]

  inputs = tokenizer.apply_chat_template(
      messages,
      tokenize = True,
      add_generation_prompt = True,
      return_tensors = "pt",
  ).to("cuda")

  outputs = model.generate(input_ids=inputs, max_new_tokens = 64, use_cache = True, temperature = 1.5, min_p = 0.1)
  pred = tokenizer.batch_decode(outputs)[0]
  target = row["text"]
  row["bleu"] = bleu(target, pred)

  metric_bleu = bleu(target, pred)
  sum_init_bleu_score += metric_bleu.item()
  metric_meteor = meteor.compute(predictions=[pred], references=[target])
  sum_init_meteor_score += metric_meteor['meteor']

  print("\navg bleu:", sum_init_bleu_score * 1. / len(eval_ds))
  print("\navg meteor:", sum_init_meteor_score * 1. / len(eval_ds))

  1%|          | 1/100 [00:06<10:59,  6.66s/it]


avg bleu: 0.006688888669013977

avg meteor: 0.003929271507737398


  2%|▏         | 2/100 [00:11<08:49,  5.40s/it]


avg bleu: 0.012807351946830749

avg meteor: 0.00955927498841961


  3%|▎         | 3/100 [00:15<07:38,  4.72s/it]


avg bleu: 0.018991196155548097

avg meteor: 0.014420192512071953


  4%|▍         | 4/100 [00:19<07:22,  4.61s/it]


avg bleu: 0.024330406188964843

avg meteor: 0.017974265168815316


  5%|▌         | 5/100 [00:23<06:55,  4.37s/it]


avg bleu: 0.030385581851005553

avg meteor: 0.02265495174491452


  6%|▌         | 6/100 [00:27<06:36,  4.22s/it]


avg bleu: 0.03693998575210571

avg meteor: 0.027101838791011815


  7%|▋         | 7/100 [00:31<06:40,  4.30s/it]


avg bleu: 0.04265044152736664

avg meteor: 0.030356160227151433


  8%|▊         | 8/100 [00:35<06:26,  4.20s/it]


avg bleu: 0.04926675915718079

avg meteor: 0.03489265995977006


  9%|▉         | 9/100 [00:39<06:15,  4.13s/it]


avg bleu: 0.05519067227840423

avg meteor: 0.03889526791426287


 10%|█         | 10/100 [00:44<06:22,  4.25s/it]


avg bleu: 0.061630581617355344

avg meteor: 0.04389567619511606


 11%|█         | 11/100 [00:48<06:10,  4.16s/it]


avg bleu: 0.06727923035621643

avg meteor: 0.04818502678496277


 12%|█▏        | 12/100 [00:52<06:01,  4.11s/it]


avg bleu: 0.07302726209163665

avg meteor: 0.05185292954500814


 13%|█▎        | 13/100 [00:56<06:07,  4.22s/it]


avg bleu: 0.07865884125232697

avg meteor: 0.05588908920054253


 14%|█▍        | 14/100 [01:00<05:57,  4.15s/it]


avg bleu: 0.0848406594991684

avg meteor: 0.06108721124408082


 15%|█▌        | 15/100 [01:04<05:51,  4.13s/it]


avg bleu: 0.0907249790430069

avg meteor: 0.06453326099110775


 16%|█▌        | 16/100 [01:09<05:54,  4.22s/it]


avg bleu: 0.09694461464881897

avg meteor: 0.06822759351044026


 17%|█▋        | 17/100 [01:13<05:44,  4.15s/it]


avg bleu: 0.10229826748371124

avg meteor: 0.07261519316110833


 18%|█▊        | 18/100 [01:17<05:41,  4.16s/it]


avg bleu: 0.1081486064195633

avg meteor: 0.07745550474825535


 19%|█▉        | 19/100 [01:21<05:40,  4.21s/it]


avg bleu: 0.1141129982471466

avg meteor: 0.08032140717427509


 20%|██        | 20/100 [01:25<05:31,  4.15s/it]


avg bleu: 0.1198231190443039

avg meteor: 0.08492042177608465


 21%|██        | 21/100 [01:30<05:30,  4.19s/it]


avg bleu: 0.12545267939567567

avg meteor: 0.08893289762868553


 22%|██▏       | 22/100 [01:34<05:28,  4.21s/it]


avg bleu: 0.13145818889141084

avg meteor: 0.09375015768284259


 23%|██▎       | 23/100 [01:38<05:20,  4.16s/it]


avg bleu: 0.13748414278030396

avg meteor: 0.09779524066057668


 24%|██▍       | 24/100 [01:42<05:22,  4.24s/it]


avg bleu: 0.14362242877483367

avg meteor: 0.10291954125432921


 25%|██▌       | 25/100 [01:46<05:14,  4.20s/it]


avg bleu: 0.14960557460784912

avg meteor: 0.10685925780926082


 26%|██▌       | 26/100 [01:50<05:06,  4.15s/it]


avg bleu: 0.155378697514534

avg meteor: 0.11096564084958216


 27%|██▋       | 27/100 [01:55<05:11,  4.26s/it]


avg bleu: 0.1617269867658615

avg meteor: 0.11454488796039378


 28%|██▊       | 28/100 [01:59<05:02,  4.20s/it]


avg bleu: 0.16688897728919982

avg meteor: 0.11858786160213665


 29%|██▉       | 29/100 [02:03<04:55,  4.16s/it]


avg bleu: 0.17250777244567872

avg meteor: 0.12343883479258957


 30%|███       | 30/100 [02:08<04:58,  4.27s/it]


avg bleu: 0.17823783218860625

avg meteor: 0.12727757101325943


 31%|███       | 31/100 [02:12<04:51,  4.23s/it]


avg bleu: 0.18466005444526673

avg meteor: 0.13161453441114715


 32%|███▏      | 32/100 [02:16<04:45,  4.19s/it]


avg bleu: 0.19143350958824157

avg meteor: 0.13599653773439713


 33%|███▎      | 33/100 [02:20<04:48,  4.31s/it]


avg bleu: 0.19731263041496278

avg meteor: 0.14026788299684323


 34%|███▍      | 34/100 [02:24<04:39,  4.23s/it]


avg bleu: 0.20302691638469697

avg meteor: 0.14436161463036257


 35%|███▌      | 35/100 [02:29<04:32,  4.19s/it]


avg bleu: 0.2091602498292923

avg meteor: 0.14875447928562172


 36%|███▌      | 36/100 [02:34<04:49,  4.52s/it]


avg bleu: 0.21463137805461885

avg meteor: 0.15357461545603726


 37%|███▋      | 37/100 [02:38<04:37,  4.40s/it]


avg bleu: 0.22103704273700714

avg meteor: 0.15923855740909992


 38%|███▊      | 38/100 [02:42<04:27,  4.32s/it]


avg bleu: 0.22691780149936677

avg meteor: 0.16374902185864124


 39%|███▉      | 39/100 [02:47<04:28,  4.40s/it]


avg bleu: 0.23262810766696929

avg meteor: 0.1687830891703907


 40%|████      | 40/100 [02:51<04:18,  4.31s/it]


avg bleu: 0.23823308885097505

avg meteor: 0.17384434513593605


 41%|████      | 41/100 [02:55<04:12,  4.27s/it]


avg bleu: 0.24494950652122496

avg meteor: 0.1775955602047844


 42%|████▏     | 42/100 [02:59<04:11,  4.34s/it]


avg bleu: 0.2505302429199219

avg meteor: 0.18176004509600946


 43%|████▎     | 43/100 [03:04<04:03,  4.27s/it]


avg bleu: 0.25694533705711364

avg meteor: 0.18530340391173705


 44%|████▍     | 44/100 [03:08<03:58,  4.27s/it]


avg bleu: 0.2632986134290695

avg meteor: 0.18875502345143375


 45%|████▌     | 45/100 [03:12<03:56,  4.31s/it]


avg bleu: 0.2695849204063416

avg meteor: 0.19332839567924434


 46%|████▌     | 46/100 [03:16<03:49,  4.25s/it]


avg bleu: 0.27570309102535245

avg meteor: 0.197342971020071


 47%|████▋     | 47/100 [03:21<03:48,  4.31s/it]


avg bleu: 0.2822655910253525

avg meteor: 0.20143054498456806


 48%|████▊     | 48/100 [03:25<03:42,  4.27s/it]


avg bleu: 0.2879918479919434

avg meteor: 0.20605544888986166


 49%|████▉     | 49/100 [03:29<03:34,  4.22s/it]


avg bleu: 0.29432874619960786

avg meteor: 0.20956047459774418


 50%|█████     | 50/100 [03:34<03:35,  4.32s/it]


avg bleu: 0.30071428835392

avg meteor: 0.21397204059902272


 51%|█████     | 51/100 [03:38<03:28,  4.25s/it]


avg bleu: 0.30657224655151366

avg meteor: 0.21832583389142088


 52%|█████▏    | 52/100 [03:42<03:22,  4.21s/it]


avg bleu: 0.3133144563436508

avg meteor: 0.22190109654486417


 53%|█████▎    | 53/100 [03:46<03:23,  4.33s/it]


avg bleu: 0.31893052279949186

avg meteor: 0.22646720432929532


 54%|█████▍    | 54/100 [03:51<03:16,  4.27s/it]


avg bleu: 0.32486887395381925

avg meteor: 0.2298808031889707


 55%|█████▌    | 55/100 [03:55<03:09,  4.21s/it]


avg bleu: 0.33122990548610687

avg meteor: 0.2339217800106327


 56%|█████▌    | 56/100 [03:59<03:10,  4.34s/it]


avg bleu: 0.33741572439670564

avg meteor: 0.23761163475032368


 57%|█████▋    | 57/100 [04:03<03:03,  4.26s/it]


avg bleu: 0.34387286722660065

avg meteor: 0.2420753248099539


 58%|█████▊    | 58/100 [04:07<02:57,  4.22s/it]


avg bleu: 0.350255845785141

avg meteor: 0.24710125857275864


 59%|█████▉    | 59/100 [04:12<02:58,  4.36s/it]


avg bleu: 0.3564822608232498

avg meteor: 0.25146483353647875


 60%|██████    | 60/100 [04:16<02:51,  4.29s/it]


avg bleu: 0.36197038769721984

avg meteor: 0.2561135352351047


 61%|██████    | 61/100 [04:21<02:47,  4.29s/it]


avg bleu: 0.36842372119426725

avg meteor: 0.26058311283867935


 62%|██████▏   | 62/100 [04:25<02:44,  4.34s/it]


avg bleu: 0.37449607253074646

avg meteor: 0.26440551401160167


 63%|██████▎   | 63/100 [04:29<02:38,  4.29s/it]


avg bleu: 0.3810673314332962

avg meteor: 0.26848436265766945


 64%|██████▍   | 64/100 [04:34<02:35,  4.31s/it]


avg bleu: 0.3864273315668106

avg meteor: 0.2736187099237479


 65%|██████▌   | 65/100 [04:38<02:30,  4.30s/it]


avg bleu: 0.3918990099430084

avg meteor: 0.2779441330358736


 66%|██████▌   | 66/100 [04:42<02:24,  4.25s/it]


avg bleu: 0.3980029058456421

avg meteor: 0.2825540912171507


 67%|██████▋   | 67/100 [04:47<02:23,  4.36s/it]


avg bleu: 0.4043405681848526

avg meteor: 0.2867764692754334


 68%|██████▊   | 68/100 [04:51<02:17,  4.29s/it]


avg bleu: 0.41119623839855196

avg meteor: 0.29095060406785744


 69%|██████▉   | 69/100 [04:55<02:11,  4.24s/it]


avg bleu: 0.4171336418390274

avg meteor: 0.29487711265735983


 70%|███████   | 70/100 [04:59<02:10,  4.36s/it]


avg bleu: 0.42375018298625944

avg meteor: 0.29881559845782485


 71%|███████   | 71/100 [05:04<02:04,  4.29s/it]


avg bleu: 0.4298508924245834

avg meteor: 0.30232978862654153


 72%|███████▏  | 72/100 [05:08<01:58,  4.24s/it]


avg bleu: 0.43526753783226013

avg meteor: 0.30837502805414047


 73%|███████▎  | 73/100 [05:12<01:57,  4.35s/it]


avg bleu: 0.44170459270477297

avg meteor: 0.3124682843290411


 74%|███████▍  | 74/100 [05:16<01:51,  4.28s/it]


avg bleu: 0.44756020784378053

avg meteor: 0.316508346066003


 75%|███████▌  | 75/100 [05:21<01:46,  4.24s/it]


avg bleu: 0.4533379858732223

avg meteor: 0.3206974797026509


 76%|███████▌  | 76/100 [05:25<01:44,  4.33s/it]


avg bleu: 0.4601507550477982

avg meteor: 0.32532466402932214


 77%|███████▋  | 77/100 [05:29<01:38,  4.28s/it]


avg bleu: 0.4664717429876328

avg meteor: 0.3300591688941891


 78%|███████▊  | 78/100 [05:34<01:34,  4.28s/it]


avg bleu: 0.4729765021800995

avg meteor: 0.33439374294742025


 79%|███████▉  | 79/100 [05:38<01:30,  4.32s/it]


avg bleu: 0.47862470388412476

avg meteor: 0.3388195544173025


 80%|████████  | 80/100 [05:42<01:25,  4.27s/it]


avg bleu: 0.48453192234039305

avg meteor: 0.3420546980586743


 81%|████████  | 81/100 [05:47<01:22,  4.32s/it]


avg bleu: 0.49062782526016235

avg meteor: 0.3480238842755869


 82%|████████▏ | 82/100 [05:51<01:17,  4.29s/it]


avg bleu: 0.4964078766107559

avg meteor: 0.3530514927956079


 83%|████████▎ | 83/100 [05:55<01:12,  4.24s/it]


avg bleu: 0.5023923325538635

avg meteor: 0.3571930047058382


 84%|████████▍ | 84/100 [05:59<01:09,  4.34s/it]


avg bleu: 0.5077440142631531

avg meteor: 0.36059757497135764


 85%|████████▌ | 85/100 [06:04<01:04,  4.27s/it]


avg bleu: 0.5141461771726609

avg meteor: 0.36462006626771576


 86%|████████▌ | 86/100 [06:08<00:59,  4.23s/it]


avg bleu: 0.5201883935928344

avg meteor: 0.3689712682204735


 87%|████████▋ | 87/100 [06:12<00:56,  4.35s/it]


avg bleu: 0.526702750325203

avg meteor: 0.3731818486910767


 88%|████████▊ | 88/100 [06:16<00:51,  4.29s/it]


avg bleu: 0.5329027503728867

avg meteor: 0.3771575135612599


 89%|████████▉ | 89/100 [06:21<00:46,  4.24s/it]


avg bleu: 0.5394553178548813

avg meteor: 0.3818008851354949


 90%|█████████ | 90/100 [06:25<00:43,  4.34s/it]


avg bleu: 0.5460598814487457

avg meteor: 0.3856991902202407


 91%|█████████ | 91/100 [06:29<00:38,  4.28s/it]


avg bleu: 0.5525356525182724

avg meteor: 0.3902100452536027


 92%|█████████▏| 92/100 [06:33<00:33,  4.23s/it]


avg bleu: 0.5587105160951614

avg meteor: 0.3946402583158212


 93%|█████████▎| 93/100 [06:38<00:30,  4.33s/it]


avg bleu: 0.5639399033784867

avg meteor: 0.3979781532405549


 94%|█████████▍| 94/100 [06:42<00:25,  4.28s/it]


avg bleu: 0.5700751692056656

avg meteor: 0.4026363861722818


 95%|█████████▌| 95/100 [06:46<00:21,  4.28s/it]


avg bleu: 0.5758501690626144

avg meteor: 0.40728714800066995


 96%|█████████▌| 96/100 [06:51<00:17,  4.33s/it]


avg bleu: 0.5817264574766159

avg meteor: 0.4110047037232382


 97%|█████████▋| 97/100 [06:55<00:12,  4.28s/it]


avg bleu: 0.5862440013885498

avg meteor: 0.41411019426539464


 98%|█████████▊| 98/100 [06:59<00:08,  4.31s/it]


avg bleu: 0.5918604397773742

avg meteor: 0.4184909603105991


 99%|█████████▉| 99/100 [07:04<00:04,  4.31s/it]


avg bleu: 0.5974792516231537

avg meteor: 0.423177405410361


100%|██████████| 100/100 [07:08<00:00,  4.28s/it]


avg bleu: 0.60386814057827

avg meteor: 0.4276240533227338


# Таким образом, результаты следующие

avg bleu: 0.57 -> 0.60

avg meteor: 0.40 -> 0.42

Разница невиликая, но пробный ответ заметно лучше